src/ucs/sys/sock.c

```diff
@@ -951,6 +956,87 @@ int ucs_sockaddr_ip_cmp(const struct sockaddr *sa1, const struct sockaddr *sa2)
                   UCS_IPV4_ADDR_LEN : UCS_IPV6_ADDR_LEN);
 }
 
+static void ucs_sockaddr_log_subnet_info(const struct sockaddr *sa1,
+                                         const struct sockaddr *sa2,
+                                         unsigned prefix_len, int matched)
+{
+    UCS_STRING_BUFFER_ONSTACK(info, UCS_SOCKADDR_STRING_LEN * 2 + 60);
+    const struct sockaddr *saddrs[] = {sa1, sa2};
+    unsigned sa_index               = 0;
+    char ip_str[UCS_SOCKADDR_STRING_LEN];
+    const struct sockaddr **sa;
+    ucs_status_t status;
+
+    ucs_string_buffer_appendf(&info, "IP addresses");
+    if (!matched) {
+        ucs_string_buffer_appendf(&info, " do not");
+    }
+
+    ucs_string_buffer_appendf(&info, " match with a %u-bit prefix,",
+                              prefix_len);
+
+    ucs_carray_for_each(sa, saddrs, ucs_static_array_size(saddrs)) {
+        status = ucs_sockaddr_get_ipstr(*sa, ip_str, UCS_SOCKADDR_STRING_LEN);
+        ucs_string_buffer_appendf(&info, " sa%u: %s", ++sa_index,
+                                  (status == UCS_OK) ? ip_str : "<null>");
+    }
+
+    ucs_debug("%s", ucs_string_buffer_cstr(&info));
+}
+
+ucs_status_t ucs_sockaddr_is_same_subnet(const struct sockaddr *sa1,
+                                         const struct sockaddr *sa2,
+                                         unsigned prefix_len, int *is_same_p)
+{
+    int is_same         = 0;
+    ucs_status_t status = UCS_OK;
+    const void *ipaddr1, *ipaddr2;
+    size_t addr_size, addr_size_bits;
+
+    if (!ucs_sockaddr_check_known_af(sa1) ||
+        !ucs_sockaddr_check_known_af(sa2)) {
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
+
+    if (sa1->sa_family != sa2->sa_family) {
```

Comment:
we can first check sa_family is the same, and that ucs_sockaddr_inet_addr_sizeof is OK. after checking that, no need to check ucs_sockaddr_check_known_af any more.